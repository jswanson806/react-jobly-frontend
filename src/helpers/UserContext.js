// Import the necessary modules from React
import React, { createContext, useState } from 'react';

// Define the initial state for the user
const INITIAL_USER_STATE = {};

// Create a new context for the user
export const UserContext = createContext();

// Create a provider component for the user context
export const UserProvider = ({ children }) => {
  // Set up state to store the user object
  const [user, setUser] = useState(INITIAL_USER_STATE);

  // Render the provider component and pass the user object and setter to the context
  return (
    <UserContext.Provider value={{ user, setUser }}>
      {children}
    </UserContext.Provider>
  );
};

