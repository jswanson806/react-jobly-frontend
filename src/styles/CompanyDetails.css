.Company-detail {
    margin: 0 auto;
    max-width: 600px;
    padding: 40px;
    padding-top: 0;
    background-color: #F5E7E1;
    border-radius: 10px;
    box-shadow: 0 8px 20px rgba(0, 0, 0, 0.15);
    font-family: 'Raleway', sans-serif;
  }

  .Company-detail .Job-title {
    padding-top: 10px;
  }
  
  .Company-detail-title {
    text-align: center;
    font-size: 3rem;
    color: #618985;
  }

  .Company-detail .Job-apply-button {
    margin-left: 80%;
  }


