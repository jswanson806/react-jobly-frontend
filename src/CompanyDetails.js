import React, { useContext, useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import JoblyApi from "./api.js";
import JobCard from './JobCard.js';
import './styles/CompanyDetails.css';
import { UserContext } from "./helpers/UserContext.js";

/**
 * CompanyDetails Component
 *
 * Renders the details of a specific company, including its name and associated job cards.
 * Requires the `UserContext` to be provided by a parent component.
 * Retrieves the company information and jobs from the server using `JoblyApi` functions.
 * Uses the `useParams` hook from React Router to get the `handle` parameter from the URL.
 */


const CompanyDetails = () => {
    const {user} = useContext(UserContext);
    const [company, setCompany] = useState(null);
    const {handle} = useParams();

    useEffect(() => {
        async function getCompany() {
            let company = await JoblyApi.getCompany(handle);
            setCompany(company);
        }
        getCompany()
    }, [handle])

    let jobs = null;

    // true when getCompany() resolves and company.jobs has a value
    if(company && company.jobs){
        // map each job to a job card
        jobs = company.jobs.map(job => (
            <JobCard job={job} key={job.title} user={user}/>
        ))
    }
    

    return(
        <div className="Company-detail" data-testid="Company-detail">
            <div className="Company-detail-heading" data-testid="Company-detail-heading">
                <h4 className="Company-detail-title" data-testid="Company-detail-title">{company && company.handle}</h4>
            </div>
            <div className="Company-detail-cards" data-testid="Company-detail-cards">
                {jobs}
            </div>
        </div>
    );
};

export default CompanyDetails;