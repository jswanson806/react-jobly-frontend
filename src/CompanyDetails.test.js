import React from "react";
import CompanyDetails from "./CompanyDetails.js";
import { act, render, waitFor, getByTestId, screen} from '@testing-library/react';
import { MemoryRouter} from 'react-router-dom';
import '@testing-library/jest-dom';
import JoblyApi from "./api.js";
import { UserContext } from "./helpers/UserContext.js";

jest.mock('./api');

const user = {
    username: "testuser",
    password: "password",
    firstName: "firstName",
    lastName: "lastName",
    email: "test@email.com",
    applications: []
}

const company = {
    handle: "test-company-1",
    name: "Test Company 1",
    num_employees: 9999,
    description: "Test description for Test Company 1",
    jobs: [
      {
        id: 9999,
        title: "Test Job Title 1",
        salary: 123456,
        equity: 0.5,
      },
    ],
}


describe('CompanyDetails smoke and snapshot', () => {

    
    it('should render without crashing', async () => {

        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                    <MemoryRouter>
                        <CompanyDetails />
                    </MemoryRouter>
                </UserContext.Provider>
            )
        })
    });


    it('should match the snapshot', async () => {
        let asFragment;
        await act(async () => {const {asFragment: fragment} = render(
            <UserContext.Provider value={{user}}>
                <MemoryRouter>
                    <CompanyDetails />
                </MemoryRouter>
            </UserContext.Provider>
            );
            asFragment = fragment;
        });
        
        expect(asFragment()).toMatchSnapshot();
    });

});

describe('Render with correct data', () => {

    it('should render JobCard components correctly', async () => {
        // mock call to api to get company
       JoblyApi.getCompany.mockReturnValueOnce(company);

        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                    <MemoryRouter>
                        <CompanyDetails />
                    </MemoryRouter>
                </UserContext.Provider>
            )
        })

        // wait for the asyncronous function call
        waitFor(() => {
            expect(getByTestId('Job-apply-button-active')).toBeInTheDocument();
        })

        // expect a single call to the api upon render from useEffect hook
        expect(JoblyApi.getCompany).toHaveBeenCalledTimes(1);

        //JobCard component has been rendered
        const jobCardButton = screen.getByTestId('Job-apply-button-active')
        expect(jobCardButton).toBeInTheDocument();

        //JobCard was only rendered once because company.jobs contains only one item
        expect(screen.getAllByTestId('Job-apply-button-active').length).toBe(1)
    });

})