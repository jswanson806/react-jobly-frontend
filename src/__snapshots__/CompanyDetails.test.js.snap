// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyDetails smoke and snapshot should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="Company-detail"
    data-testid="Company-detail"
  >
    <div
      class="Company-detail-heading"
      data-testid="Company-detail-heading"
    >
      <h4
        class="Company-detail-title"
        data-testid="Company-detail-title"
      />
    </div>
    <div
      class="Company-detail-cards"
      data-testid="Company-detail-cards"
    />
  </div>
</DocumentFragment>
`;
